prov/efa: use local read to copy data to GPU buffer

wzamazon opened the pull request:
This PR contain 12 commits that implements using local read to copy data to GPU buffer.

The first 7 commits are refactors.

The next 5 commits implements the local read.

wzamazon commented:
bot:aws:retest

wzamazon commented:
bot:aws:retest

wzamazon commented:
The single-node test failure is because of instances issue.

```
 172.31.19.101 ssh: connect to host 172.31.19.101 port 22: Connection timed out
```
The test passed before, so I am not going to retest.

wzamazon commented:
bot:aws:retest

bwbarrett commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,46 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
 	return -1;
 }
 
+ssize_t rxr_read_mr_reg(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
```
Function documentation?

wzamazon replied:
Thanks! added in the new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,46 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+				FI_RECV, 0, 0, 0, &read_entry->mr[i], NULL);
+
+		if (err) {
+			/* In case error happened, we did not close registered MR here, because:
```
then where does the MR get closed?

wzamazon replied:
They are closed in ```rxr_read_release_entry```,  I updated comments in new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,46 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+
+		if (err) {
+			/* In case error happened, we did not close registered MR here, because:
+			 * if error is -FI_ENOMEM, we return -FI_EAGAIN, and the registration will
```
I can read the if statement, but I can't get the why from the code.  Why are we doing this kind of round about behavior?

wzamazon replied:
We return -FI_EAGAIN when encountered -FI_ENOMEM, because progress engine will close registrations and try again.

bwbarrett commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,46 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+	return 0;
+}
+
 struct rxr_read_entry *rxr_read_alloc_entry(struct rxr_ep *ep, int entry_type, void *x_entry,
```
If you update a function, please take time to add documentation comments.

wzamazon replied:
Thanks! added in the new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,70 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
 	return 0;
 }
 
+static inline
```
function documentation

wzamazon replied:
added in new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,70 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+		break;
+	case RXR_EAGER_MSGRTM_PKT:
+	case RXR_EAGER_TAGRTM_PKT:
+		rxr_pkt_handle_eager_rtm_data_copied(ep, pkt_entry);
```
Splitting out all the rtm data copied cases feels ugly.  Also, still feels like we have the state machine steps wrong here and are treating this as a special state machine off to the side.

wzamazon replied:
Yes, you are right. I was mixed "data received" state and "data copied" state. CTS packet only depends on data being received. writing completion and releasing rx entry depend on data being copied. 
The new revision split the two state, therefore this function ```rxr_pkt_handle_data_copied``` is much simpler.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,70 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"invalid control pkt type %d\n",
+			rxr_get_base_hdr(pkt_entry->pkt)->type);
+		assert(0 && "invalid control pkt type");
```
why do you have an assert and further error processing?

wzamazon replied:
removed in new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,70 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+	}
+}
+
+void rxr_pkt_proc_data(struct rxr_ep *ep,
```
Function documentation?  Also, there's no verb in the function name; I'm not sure exactly what this function does from the name (other than take an obscene number of arguments).

wzamazon replied:
renamed this function to ```rxr_pkt_copy_to_rx()``` in new revision.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,70 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+	}
+
+	desc = rx_entry->desc[0];
+	ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
```
I'm surprised in a function named `rxr_pkt_proc_data()` to see a blind call to `ofi_copy_to_hmem_iov()`.  Something is not as descriptive as it should be.

wzamazon replied:
The new function name is ```rxr_pkt_copy_to_rx``` thus does not have this problem.

bwbarrett commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1149,6 +1152,7 @@ static void rxr_buf_region_free_hndlr(struct ofi_bufpool_region *region)
 
 static int rxr_create_pkt_pool(struct rxr_ep *ep, size_t size,
 			       size_t chunk_count,
+			       size_t flags,
```
why isn't this change in its own patch?

wzamazon replied:
splitter this into a separate patch.

bwbarrett commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1308,6 +1336,9 @@ int rxr_ep_init(struct rxr_ep *ep)
 err_free_tx_entry_pool:
 	if (ep->tx_entry_pool)
 		ofi_bufpool_destroy(ep->tx_entry_pool);
+err_free_rx_readcopy_pool:
```
This whole chain of labels is overkill.  Why have multiple jump points, if you are also going to conditional test before free?

wzamazon replied:
I agree, but it probably is better to make this change in a separate commit.

bwbarrett commented on prov/efa/src/rxr/rxr_pkt_entry.c:
```diff
@@ -182,6 +182,12 @@ void rxr_pkt_entry_release_rx(struct rxr_ep *ep,
 			ep->rx_bufs_efa_to_post++;
 	}
+
+	if (pkt_entry->type == RXR_PKT_ENTRY_READ_COPY) {
```
there are a lot of if statements in here; is this really a good path to continue for something that is in the critical path?

wzamazon replied:
Unfortunately, the caller of this function does not know the ```pkt_entry->type``` either, so there has to be a branch. This branch is just for monitoring ```rx_readcopy_pkt_pool``` usage, maybe I can put it in debug?

bwbarrett commented on prov/efa/src/rxr/rxr.h:
```diff
@@ -226,6 +226,7 @@ enum rxr_lower_ep_type {
 enum rxr_x_entry_type {
 	RXR_TX_ENTRY = 1,
 	RXR_RX_ENTRY,
+	RXR_PKT_ENTRY,
```
this name is not very descriptive.  Let's try better, with some documentation.

wzamazon replied:
I new revision, I created a new enum ```rxr_read_context_type``` and added more comments about each type context.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -82,7 +82,47 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+		if (!pkt_entry_copy) {
+			FI_WARN(&rxr_prov, FI_LOG_CQ,
+				"readcopy pkt pool exhausted! Set FI_EFA_READCOPY_POOL_SIZE to a higher value!");
+			return -FI_ENOBUFS;
```
I thought we were going to queue in this case? This will write an error (and crash if no EQ).

wzamazon replied:
Yes, changed to -FI_EAGAIN in new revision.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -82,7 +82,47 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_pkt_entry *pkt_entry_copy;
+
+	if (read_entry->x_entry_type == RXR_PKT_ENTRY) {
```
This section doesn't seem applicable for a function called `rxr_read_mr_reg`, can this go in its own function?

wzamazon replied:
This section is still preparing memory registration, so I think it is appropriate to leave it here. Maybe rename the function to ```rxr_read_prepare_mr```?

rwespetal replied:
I'd still split the function even if it has to be called in `rxr_read_mr_reg` for readability.

wzamazon replied:
Split into a new function in new revision.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -82,7 +82,47 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+		read_entry->rma_iov[0].addr = (uint64_t)pkt_entry_copy->pkt + pkt_offset;
+		read_entry->rma_iov[0].key = fi_mr_key(pkt_entry_copy->mr);
+
+		/* in this case, iov is on GPU buffer and is always registered, so no need to go furthure
```
typo here

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -137,10 +177,10 @@ struct rxr_read_entry *rxr_read_alloc_entry(struct rxr_ep *ep, int entry_type, v
 
 		read_entry->iov_count = tx_entry->iov_count;
-		read_entry->iov = tx_entry->iov;
+		memcpy(read_entry->iov, tx_entry->iov, tx_entry->iov_count * sizeof(struct iovec)) ;
```
Is this (and the next three memcpy's added) related to this change? Or is this a separate bug?

wzamazon replied:
It is not a bug, but a related change. Previously, we use pointer for ```iov``` and ```rma_iov```. They are just pointing to ```iov``` and ```rma_iov``` in tx_entry and rx_entry.

With this change, we have to construct ```iov``` and ```rma_iov``` on the fly, so I change them from pointer to array.

I added a separate PR for this change in new revision.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -235,18 +294,67 @@ int rxr_read_post_or_queue(struct rxr_ep *ep, int entry_type, void *x_entry)
+		return -FI_ENOBUFS;
+	}
+
+	assert(ep->rdm_self_addr != FI_ADDR_NOTAVAIL);
```
This is duplicating code that's already in the rx/tx entry init and handling logic. It would be much cleaner if we can avoid the `rxr_read_entry` structure but it seems like you're using it to create another iov to consume, right? We should rethink this later.

wzamazon replied:
Yes, I think eventually we want to merge tx_entry, rx_entry and read_entry into one structure.

wzamazon commented on prov/efa/src/rxr/rxr_read.h:
```diff
@@ -53,7 +53,8 @@ struct rxr_read_entry {
 
 	enum rxr_x_entry_type x_entry_type;
-	int x_entry_id;
+	void *x_entry;
```
added in the new revision.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,46 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+				FI_RECV, 0, 0, 0, &read_entry->mr[i], NULL);
+
+		if (err) {
+			/* In case error happened, we did not close registered MR here, because:
```
sorry, the comment is a little misleading. What I meant to say is: in case of multiple iov, if the i-th registration failed, ```mr[0]``` to ```mr[i-1]``` are still registered, and need to be closed eventually. We do not close them here, because they will be closed in ```rxr_read_release_entry```.

rwespetal replied:
Thank you, this makes sense to me now.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -129,47 +165,20 @@ struct rxr_read_entry *rxr_read_alloc_entry(struct rxr_ep *ep, int entry_type, v
-						read_entry->iov[i].iov_base, read_entry->iov[i].iov_len,
-						FI_RECV, 0, 0, 0, &read_entry->mr[i], NULL);
-
-				if (err == -FI_ENOMEM && efa_mr_cache_enable) {
```
Before this change, registration happened during initialization and the retry is explicit.

This change move the registration to ```rxr_read_post```, and will return -FI_EAGAIN when registration failed. When progress engine get -FI_EAGAIN, it will retry, so we do not need to explicitly retry any more. I updated commit message to discuss this.

wzamazon commented:
aws ci failure is because of instance failure
```
ssh: connect to host 172.31.24.92 port 22: Connection timed out
```

wzamazon commented:
AWS ci was because instance cannot connect to GitHub.  retesting ....

wzamazon commented:
bot:aws:retest

rwespetal commented on prov/efa/src/rxr/rxr_read.h:
```diff
@@ -38,22 +38,43 @@
+ * 2. read message protocol is being used, receiver is going
+ *    to post a read requst.
+ *
+ * To distinguish them, we use a pointer as context.
```
This isn't needed, there is already a type field in rxr_tx_entry/rxr_rx_entry.

wzamazon replied:
Yes, it is not needed now. This is mainly preparation for the future when we are adding a case packet entry used as context. Because ```rxr_pkt_entry``` has to remain 64 bytes, we cannot add any new members to it.

rwespetal replied:
I see - I hadn't read the last commit before making this comment.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,67 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+			 * The progress engine will release registrations and call this
+			 * function again.
+			 *
+			 * All registration opened here will be closed during release of
```
Is this correct? Looks like it leaves the registrations alone and queues the message.

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -79,15 +79,67 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
+
+		if (err) {
+			/* If registration failed with -FI_ENOMEM, we return -FI_EAGAIN.
+			 * The progress engine will release registrations and call this
```
Is this correct? Looks like the progress engine leaves the registrations alone and queues the message when FI_EAGAIN is returned.

wzamazon replied:
Sorry, the comment is a little misleading. I meant to say "the progress engine will progress other message transfers, which will release registrations. Therefore, when the progress engine call this function again later, there will be registrations available".

rwespetal commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -709,8 +709,14 @@ static void rxr_ep_free_res(struct rxr_ep *rxr_ep)
 
-	if (rxr_ep->rx_readcopy_pkt_pool)
+	if (rxr_ep->rx_readcopy_pkt_pool) {
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "current usage of read copy packet pool is %d\n",
```
I'd change these to FI_INFO or FI_DBG not warn

wzamazon replied:
changed to FI_INFO in new revision.

rwespetal reviewed (commented):
Finished reviewing everything but the last commit `prov/efa: use local read to copy data to gpu receiving buffer`

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_req.c:
```diff
@@ -1372,23 +1318,8 @@ void rxr_pkt_handle_eager_rtw_recv(struct rxr_ep *ep,
-		return;
-	}
-
-	if (rx_entry->cq_entry.flags & FI_REMOTE_CQ_DATA)
```
The FI_REMOTE_CQ_DATA logic is handled by ```rxr_cq_handle_rx_completion``` and is tested by the rma writedata fabtest.

I will add back the logic to check ```bytes_left```.

rwespetal replied:
I see, thank you.

wzamazon replied:
added in new revision.

rwespetal commented:
I made it through the series and I'm fine with the changes once the bytes_left check is re added.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_cmd.c:
```diff
@@ -449,6 +450,41 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+				     data_size);
+	}
+
+	rx_entry->bytes_done += data_size;
```
added the check back in new revision, also added the validations back

wzamazon commented:
@rwespetal I believe I addressed all your comments in new revision, including:

 (1) check for ```bytes_left```
(2) a separate function for pkt_entry mr
(3) FI_WARN -> FI_INFO for readcopy_pkt_pool size, let me know if I missed anything.

rwespetal commented:
bot:aws:retest

rwespetal commented:
NCCL tests continue to hit intermittent failures that are unrelated to this PR and are being tracked elsewhere.
